src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -0,0 +1,213 @@
+    }
+}
+
+/**
+ * The DebuggingManager maintains the mapping between notebook documents and debug sessions.
+ */
+@injectable()
+export class DebuggingManager implements IExtensionSingleActivationService {
+    private notebookToDebugger = new Map<vscode.NotebookDocument, Debugger>();
+    private readonly disposables: IDisposable[] = [];
+
+    public constructor(
+        @inject(IKernelProvider) private kernelProvider: IKernelProvider,
+        @inject(INotebookProvider) private notebookProvider: INotebookProvider,
+        @inject(IDebuggingCellMap) private debuggingCellMap: IDebuggingCellMap,
+        @inject(INotebookControllerManager) private readonly notebookControllerManager: INotebookControllerManager
+    ) {}
+
+    public async activate() {
+        vscode.debug.breakpoints; // start to fetch breakpoints
+
+        this.disposables.push(
+            // track termination of debug sessions
+            vscode.debug.onDidTerminateDebugSession(async (session) => {
+                for (const [doc, dbg] of this.notebookToDebugger.entries()) {
+                    if (dbg && session === (await dbg.session)) {
+                        this.debuggingCellMap.getCellsAnClearQueue(doc);
+                        this.notebookToDebugger.delete(doc);
+                        break;
+                    }
+                }
+            }),
+
+            // track closing of notebooks documents
+            vscode.workspace.onDidCloseNotebookDocument(async (document) => {
+                this.debuggingCellMap.getCellsAnClearQueue(document);
+                const dbg = this.notebookToDebugger.get(document);
+                if (dbg) {
+                    await dbg.stop();
+                }
+                this.fixBreakpoints(document);
+            }),
+
+            // factory for kernel debug adapters
+            vscode.debug.registerDebugAdapterDescriptorFactory('kernel', {
+                createDebugAdapterDescriptor: async (session) => {
+                    const activeDoc = vscode.window.activeNotebookEditor!.document;
+
+                    await this.ensureKernelIsRunning(activeDoc);
+                    const debug = this.getDebuggerByUri(activeDoc);
+
+                    if (debug) {
+                        const notebook = await this.notebookProvider.getOrCreateNotebook({
+                            resource: debug.document.uri,
+                            identity: debug.document.uri,
+                            getOnly: true
+                        });
+                        if (notebook && notebook.session) {
+                            debug.resolve(session);
+                            return new vscode.DebugAdapterInlineImplementation(
+                                new KernelDebugAdapter(session, debug.document, notebook.session, this.debuggingCellMap)
+                            );
+                        } else {
+                            void vscode.window.showInformationMessage('run the kernel');
```

Comment:
Might be good to mark spots like this (assuming this is a temp error message) with TODOs so you can add localized messages before final PRs.

Reply:
Also we should use our specific messaging classes.

Reply:
cleanup commit